### ISSUE TITLE
`maxBufferSize` questions

### ISSUE THREAD

**Issue by teoxoy** [gpuweb/gpuweb#3700]
- for Vulkan, I believe we don't want to require `VK_KHR_maintenance3` to be able to query [`maxMemoryAllocationSize`](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceMaintenance3Properties.html)
  So, what limit should we go with for Vulkan without `VK_KHR_maintenance3`? the default 256MB? @kainino0x brought this up previously in https://github.com/gpuweb/gpuweb/issues/1371#issuecomment-1259923913.
- for dx11/12, I can't find any limits that translate to our `maxBufferSize`
- could we increase the default (currently at 256MB) since it feels quite [limiting](https://github.com/gpuweb/gpuweb/issues/1371#issuecomment-767319435)?
  This requires macOS 10.14+ (see also https://github.com/gpuweb/gpuweb/issues/3685) due to Metal's [`maxBufferLength`](https://developer.apple.com/documentation/metal/mtldevice/2966563-maxbufferlength?language=objc) only being available starting with 10.14.
  According to @litherum in https://github.com/gpuweb/gpuweb/issues/1371#issuecomment-1054937312 all values are at least 1GB.

**Comment by Kangz** [comment]
- For Vulkan I think the final idea was that we could make up a number and return an OOM if the allocation fails in the driver because it is too big.
 - D3D12 doesn't have a limit and I think we were told we can assume 2GB.
 - You could choose to increase the default to 1GB in your implementation, but Chromium supports WebGPU on macOS 10.13 so we can't bump it.

**Comment by kainino0x** [comment]
Yup, we definitely don't expect to expose a limit of 256MB except on the Metal devices that require it. That's a very low limit and we fully expect larger allocations will work fine on most Vulkan/D3D12 devices. I think defaulting around 1-4 GB in these cases is reasonable. You might be able to pick a limit based on the total VRAM size of the device (e.g. on android you might be very memory-limited and you wouldn't want to expose a maxBufferSize larger than the total amount of memory in the system).

**Comment by teoxoy** [comment]
My suspicion (based on [litherium's comment](https://github.com/gpuweb/gpuweb/issues/1371#issuecomment-1054937312)) is that if we agree to support macOS 10.14+ which has support for Metal's `maxBufferLength` we can in turn guarantee in the spec that the adapter/device support at least 1GB.

It's a bit sad to see such a low limit only due to a range of versions of a platform of one of the APIs we target.

**Comment by kainino0x** [comment]
We could do this now, but we still can raise the core limit after V1 as well. For now, developers are free to make the decision to drop support for the few systems with maxBufferSize below 1GB or whatever, so I think it's OK if we don't figure it out urgently.
